mage_integrations/mage_integrations/sources/paystack/streams/base.py

```diff
@@ -0,0 +1,166 @@
+        return False
+
+    elif selected is not None:
+        return selected
+
+    return inclusion == 'automatic'
+
+
+class BasePaystackStream():
+    # GLOBAL PROPERTIES
+    TABLE = None
+    KEY_PROPERTIES = []
+    API_METHOD = 'GET'
+    REQUIRES = []
+
+    def __init__(self, config, state, catalog, client: PaystackClient):
+        self.config = config
+        self.state = state
+        self.catalog = catalog
+        self.client = client
+        self.substreams = []
+
+    def write_schema(self):
+        bookmark_properties = []
+        if REPLICATION_METHOD_INCREMENTAL == self.catalog.replication_method:
+            bookmark_properties = self.BOOKMARK_PROPERTIES
+        write_schema(
+            self.catalog.stream,
+            self.catalog.schema.to_dict(),
+            key_properties=self.catalog.key_properties or [],
+            bookmark_properties=bookmark_properties,
+            replication_method=self.catalog.replication_method,
+            unique_conflict_method=self.catalog.unique_conflict_method,
+            unique_constraints=self.catalog.unique_constraints,
+        )
+
+    def get_url(self):
+        """
+        Return the URL to hit for data from this stream.
+        """
+        raise RuntimeError("get_url not implemented!")
+
+    def get_abs_path(self, path):
+        return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)
+
+    # This overrides the transform_record method in the Fistown Analytics tap-framework package
+    def transform_record(self, record):
+        with singer.Transformer(integer_datetime_fmt="unix-seconds-integer-datetime-parsing") as tx:
+            metadata = {}
+
+            if self.catalog.metadata is not None:
+                metadata = singer.metadata.to_map(self.catalog.metadata)
+
+            return tx.transform(
+                record,
+                self.catalog.schema.to_dict(),
+                metadata
+            )
+
+    def get_stream_data(self, data):
+        return [self.transform_record(item) for item in data]
+
+    def load_data(self, bookmarks: Dict = None, bookmark_properties: List = None, to_date: str = None):
+        table = self.TABLE
+        api_method = self.API_METHOD
+        done = False
+        sync_interval_in_mins = 2
+        bookmark_date = None
+
+        # Attempt to get the bookmark date from the state file (if one exists and is supplied).
+        LOGGER.info('Attempting to get the most recent bookmark_date for entity {}.'.format(self.ENTITY))
+        if bookmarks and bookmark_properties:
+            bookmark_date = bookmarks.get(bookmark_properties[0])
+
+        # If there is no bookmark date, fall back to using the start date from the config file.
+        if bookmark_date is None:
+            LOGGER.info('Could not locate bookmark_date from STATE file. Falling back to start_date from config.json instead.')
+            bookmark_datetime = parse(self.config.get('start_date'))
```

Comment:
will there always be a start_date? can we have a default value?

Reply:
i was planning to make it a required value in the config

Reply:
ok i just added a default to be safe